dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,352 @@
+        systemd.service_start(f"dispatch-{event_name}.timer")
+
+    def _stop_timer(self, event_name: str) -> None:
+        """Stop a timer.
+
+        Args:
+            event_name: The event to be fired
+        """
+        # TODO: make idempotent
+        systemd.service_disable(f"dispatch-{event_name}.timer")
+        systemd.service_stop(f"dispatch-{event_name}.timer")
+
+    def dns_record_relations_data_to_zones(
+        self,
+        relation_data: list[
+            tuple[dns_record.DNSRecordRequirerData, dns_record.DNSRecordProviderData]
+        ],
+    ) -> list[models.Zone]:
+        """Return zones from all the dns_record relations data.
+
+        Args:
+            relation_data: input relation data
+
+        Returns:
+            The zones from the record_requirer_data
+        """
+        zones: dict[str, models.Zone] = {}
+        for record_requirer_data, _ in relation_data:
+            for new_zone in self.record_requirer_data_to_zones(record_requirer_data):
+                if new_zone.domain in zones:
+                    zones[new_zone.domain].entries.update(new_zone.entries)
+                else:
+                    zones[new_zone.domain] = new_zone
+        return list(zones.values())
+
+    def dns_record_relations_data_to_entries(
+        self,
+        relation_data: list[
+            tuple[dns_record.DNSRecordRequirerData, dns_record.DNSRecordProviderData]
+        ],
+    ) -> list[models.DnsEntry]:
+        """Convert DNSRecordRequirerData to a list of DnsEntry.
+
+        Args:
+            relation_data: input relation data
+        Returns:
+            A list of DnsEntry
+        """
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(models.create_dns_entry_from_requirer_entry(entry))
+        return entries
+
+    def record_requirer_data_to_zones(
+        self,
+        record_requirer_data: dns_record.DNSRecordRequirerData,
+    ) -> list[models.Zone]:
+        """Convert DNSRecordRequirerData to zone files.
+
+        Args:
+            record_requirer_data: The input DNSRecordRequirerData
+
+        Returns:
+            A list of zones
+        """
+        zones_entries: dict[str, list[models.RequirerEntry]] = {}
+        for entry in record_requirer_data.dns_entries:
+            if entry.domain not in zones_entries:
+                zones_entries[entry.domain] = []
+            zones_entries[entry.domain].append(entry)
+
+        zones: list[models.Zone] = []
+        for domain, entries in zones_entries.items():
+            zone = models.Zone(domain=domain, entries=set())
+            for entry in entries:
+                zone.entries.add(models.create_dns_entry_from_requirer_entry(entry))
+            zones.append(zone)
+        return zones
+
+    def _get_remote_relation_data(self) -> dns_record.DNSRecordProviderData | None:
```

Comment:
This method is added just to compute execution time, do we still need this ?

Reply:
Removed